File: sorting/heap.py
def left(i):
    return 2*i+1
def right(i):
    return 2*i+2
def parent(i):
    return (i-1)//2

def max_heapify(A,i, heap_size):
    l=left(i)
    r=right(i)
    if(l<heap_size and A[l]>A[i]):
        largest=l
    else:
        largest=i
    if(r<heap_size and A[r]>A[largest]):
        largest=r

    if largest!=i:
        temp=A[i]
        A[i]=A[largest]
        A[largest]=temp
        max_heapify(A,largest, heap_size)

def min_heapify(A,i, heap_size):
    l=left(i)
    r=right(i)
    if(l<heap_size and A[l]<A[i]):
        smallest=l
    else:
        smallest=i
    if(r<heap_size and A[r]<A[smallest]):
        smallest=r
    if(smallest!=i):
        temp=A[i]
        A[i]=A[smallest]
        A[smallest]=temp
        min_heapify(A,smallest, heap_size)

def build_min_heap(A):
    for i in range(len(A)//2,-1,-1):
        min_heapify(A,i,len(A))

def build_heap(A):
    for i in range(len(A)//2+1,-1,-1):
        max_heapify(A,i, len(A))

def heap_sort(A):
    heap_size=len(A)
    for i in range(len(A)-1,0,-1):
        A[i],A[0]=A[0],A[i]
        heap_size-=1
        max_heapify(A,0, heap_size)
        print(A)
#proriety queue 
def heap_maximum(A):
    return A[0]
def heap_Extract_Max(A):
    if(len(A)==0):
        return "heap underflow"
    max=A[0]
    A[0]=A[-1]
    A.pop()
    heap_size-=1
    max_heapify(A,0,len(A))
    return max

def heap_extract_min(A):
    if(len(A)==0):
        return "heap underflow"
    heap_size=len(A)
    min=A[0]
    A[0]=A[-1]
    A.pop()
    heap_size-=1
    min_heapify(A,0,len(A))
    return min

def heap_increase_key(A,i,key):
    if key<A[i]:
        return "new key is smaller than current key"
    A[i]=key
    while i>0 and A[parent(i)]<A[i]:
        A[i],A[parent(i)]=A[parent(i)],A[i]
        i=parent(i)
    
def max_heap_insert(A,key):
    heap_size+=1
    A.append(0)
    heap_increase_key(A,len(A)-1,key)


A=[15,4,2,24,5,7,8,6,5,3,3,2]
heap_size=len(A)
build_heap(A)
print(A)
# heap_sort(A)
# print(A) 

def merge_sorted_list(lists):
    n=len(lists)
    lowest=[]
    for i in range(n):
        lowest.append(lists[i][0])
    heap_size=len(lowest)
    build_heap(lowest)
    print(lowest)
    result=[]
    while heap_size>0:
        result.append(heap_extract_min(lowest))
        if len(lists[lowest.index(result[-1])])>1:
            heap_size+=1
            lowest.append(lists[lowest.index(result[-1])][1])
            heap_increase_key(lowest,heap_size-1,lists[lowest.index(result[-1])][1])
            lists[lowest.index(result[-1])].pop(0)
            
        else:
            heap_size-=1
    return result

lists=[[1,2,3],[4,5,6],[7,8,9]]
print(merge_sorted_list(lists))

File: sorting/go/heap.go
// heap implemetaion in go
package main

import "fmt"

func left(i int) int {
	return 2*i + 1
}
func right(i int) int {
	return 2*i + 2
}
func parent(i int) int {
	return (i - 1) / 2
}

func maxHeapify(arr []int, i int, heapSize int) {
	l := left(i)
	r := right(i)
	largest := i
	if l < heapSize && arr[l] > arr[i] {
		largest = l
	}
	if r < heapSize && arr[r] > arr[largest] {
		largest = r
	}
	if largest != i {
		arr[i], arr[largest] = arr[largest], arr[i]
		maxHeapify(arr, largest, heapSize)
	}

}

func buildMaxHeap(arr []int) {
	heapSize := len(arr)
	for i := len(arr)/2 - 1; i >= 0; i-- {
		maxHeapify(arr, i, heapSize)
	}
}

func heapSort(arr []int) {
	heapSize := len(arr)
	buildMaxHeap(arr)
	for i := len(arr) - 1; i >= 1; i-- {
		arr[0], arr[i] = arr[i], arr[0]
		heapSize--
		maxHeapify(arr, 0, heapSize)
	}
}

func main() {
	arr := []int{9, 9, 7, 6, 5, 4, 3, 22, 1}
	fmt.Println("Before sorting:", arr)
	heapSort(arr)
	fmt.Println("After sorting:", arr)
}
